File: Server/src/main/java/fr/nathan/mim/api/data/DataElement.java
package fr.nathan.mim.api.data;


public abstract class DataElement implements Cloneable {

    public abstract DataElement clone();

    public DataObject getAsObject()       {return (DataObject) this;}
    public DataArray getAsArray()         {return (DataArray) this;}
    public DataPrimitive getAsPrimitive() {return (DataPrimitive) this;}

    /**
     * Transforme l'objet en une chaine de caractere
     * compréhensible par un humain.
     *
     * @return Une chaine de caractere correspondant à l'objet
     */
    public abstract String toString();

}

File: Client/src/data/DataElement.cpp
//
// Created by Nathan David on 26/01/2022.
//
#include "data/DataArray.h"
#include "data/DataObject.h"
#include "data/DataPrimitive.h"
#include "data/DataElement.h"
#include "scene/SceneDao.h"


DataObject *DataElement::getAsObject() const {
    return (DataObject *) this;
}
DataArray *DataElement::getAsArray() const {
    return (DataArray *) this;
}
DataPrimitive *DataElement::getAsPrimitive() const {
    return (DataPrimitive *) this;
}
ostream &operator<<(ostream &os, const DataElement &element) {
    return os << element.toString();
}

string typeToString(Type type) {
    return enumStr[type];
}